contents/revisionControl/bestPracticesGit.md

```diff
@@ -0,0 +1,259 @@
+
+Ensure that you do this _before_
+- pushing to a remote repository [to respect the published history](#respect-published-history) and
+- performing any merges from other branches.
+
+Read more:
+- [On Sausage Making](https://sethrobertson.github.io/GitBestPractices/#sausage)
+- [Git Tools - Rewriting History](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) (great reference for how to actually do this, under _Changing Multiple Commit Messages_)
+
+---
+
+## Respect Published History
+Always avoid rewriting the public, published history unless you have a very good reason, you are very sure of what you are doing, or you will not affect anyone else.
+
+### Exceptions
+You are working on your own branch that _no one else is using_, and
+- you want to revert a commit without introducing another commit
+- you are rebasing the branch
+- you are [cleaning up the history of your branch](#hide-the-sausage-making)
+
+Read more:
+- [Rewriting History - Atlassian](https://www.atlassian.com/git/tutorials/rewriting-history)
+- [Don't change public history.](https://sethrobertson.github.io/GitBestPractices/#pubonce)
+
+---
+
+## Keep Up To Date
+### Use Remotes Effectively
+_Remotes_ refer to versions of the project you are working on that are hosted elsewhere, usually on the Internet. Remotes are very handy for managing collaboration, e.g. if you have to keep your code in sync with the `upstream` branch of the project, or if you need to pull code from someone else which may not be merged yet.
+
+You can have as many remotes as you want, each possibly being read-only or with read/write privileges.
+
+Remotes are managed using the `git remote` command.
+```bash
+# view all remotes
+git remote -v
+
+# add a remote called "upstream" that points to https://github.com/TEAMMATES/teammates
+git remote add upstream https://github.com/TEAMMATES/teammates
+
+# branch off from the master branch of the upstream repository
+git fetch upstream # get data from upstream repo
+git checkout -b your-fancy-branch upstream/master # makes a new branch off the head of upstream/master
+```
+
+Read more:
+- [Git Basics - Working with Remotes](https://git-scm.com/book/en/v2/Git-Basics-Working-with-Remotes)
+
+### Rebase versus Merging
+| You should... | when ... |
+|-------------|-|
+| merge | you created a branch to develop a feature, and now you want these changes to be inside `master` |
+| rebase | you created a branch from `master` to develop a feature, and someone else pushed a change to `master` before you finished |
+
+It is generally considered good practice to rebase your feature branch onto whatever branch you're trying to patch _before_ you make the pull request, resolving any conflicts that arise. This:
+
+- [keeps the history clear and linear](http://www.bitsnbites.eu/a-tidy-linear-git-history/).
+    - makes backtracking easier
+    - easy to follow history
+    - reverting/rolling back is much simpler
+    - you can use `git bisect` to find regressions easily
```

Comment:
You can use `git bisect` even without rebasing to master right? Cos you can git bisect on `master`, and once you find out which `merge` is buggy, then `git bisect` the `merge/PR`.